tools/rtpsrelay/RelayHandler.cpp

```diff
@@ -281,21 +281,65 @@ void HorizontalHandler::process_message(const ACE_INET_Addr&,
 
 SpdpHandler::SpdpHandler(ACE_Reactor* a_reactor,
                          GroupTable& a_group_table,
-                         RoutingTable& a_routing_table)
+                         RoutingTable& a_routing_table,
+                         const OpenDDS::DCPS::RepoId& application_participant_guid,
+                         const ACE_Time_Value& lifespan,
+                         const ACE_Time_Value& purge_period)
   : VerticalHandler(a_reactor, a_group_table, a_routing_table)
   , mutable_group_table_(a_group_table)
-{}
+  , application_participant_guid_(guid_to_string(application_participant_guid))
+  , lifespan_(lifespan)
+{
+  reactor()->schedule_timer(this, 0, purge_period, purge_period);
+}
 
 void SpdpHandler::process_message(const ACE_INET_Addr& a_remote,
                                   const ACE_Time_Value& a_now,
                                   const std::string& a_src_guid,
                                   ACE_Message_Block* a_buffer,
                                   bool is_beacon_message)
 {
+  const std::string remote_str = addr_to_string(a_remote);
+
+  const ACE_Time_Value expiration = a_now + lifespan_;
+  std::pair<AddrExpirationMap::iterator, bool> res = addr_expiration_map_.insert(std::make_pair(remote_str, expiration));
+  if (!res.second) {
+    // Exists.  Cleanup the expiration map.
+    const ACE_Time_Value previous_expiration = res.first->second;
+    std::pair<ExpirationAddrMap::iterator, ExpirationAddrMap::iterator> r = expiration_addr_map_.equal_range(previous_expiration);
+    while (r.first != r.second && r.first->second != remote_str) {
+      ++r.first;
+    }
+    expiration_addr_map_.erase(r.first);
+    // Assign the new expiration time.
+    res.first->second = expiration;
+  }
+  expiration_addr_map_.insert(std::make_pair(expiration, remote_str));
+
   if (is_beacon_message) {
     return;
   }
 
+  if (a_src_guid == application_participant_guid_) {
+    if (application_participant_addr_.empty()) {
+      application_participant_addr_ = remote_str;
+      ACE_DEBUG((LM_INFO, "(%P|%t) %N:%l SpdpHandler::handle_input application participant %s has address %s\n", application_participant_guid_.c_str(), application_participant_addr_.c_str()));
+    }
+    // Forward to all peers except the application participant.
+    for (const auto p : addr_expiration_map_) {
+      if (p.first != application_participant_addr_) {
+        enqueue_message(p.first, a_buffer);
+      }
+    }
+
+    return;
+  }
+
+  // Always send to the application participant.
```

Comment:
Does this `Always` include the cases where line 335 returns?

Reply:
No.  If the message was from the application participant it does not need to be reflected.

Reply:
I guess I just found this comment a little confusing.  It may be better without the comment, or with a different comment.  Or maybe it's just me.

These types of comments (line 328 and this one) look like they were placeholders that didn't get removed.